elastic/logstash
# live timers: API boundary, implementation, opt-in

## Release notes

 [rn:skip]


## What does this PR do?

Introduces an API boundary for timers as a first-class metric, as described in elastic/logstash#14675, and migrates all known internal timers to use the new API boundary for tracked execution.

Please refer to the specification for details on motivations.

## Why is it important/What is the impact to the user?

STUB: as of PR creation, there is no impact to the user. Developers using the Ruby- or Java-based Plugin APIs will have access to a new API boundary for acquiring a timer object that they can use to track execution time.

## Checklist



- ~[ ] I have made corresponding changes to the documentation~ _[deferred: #14809]_

## Author's Checklist


- [x] introduce write-only `TimerMetric` to  Java API
- [x] introduce read-write internal `TimerMetric` interface that extends `Metric<Long>`
- [x] implement our internal `TimerMetric` with _status-quo_ `AfterCompletionTimerMetric`
- [x] plumb `TimerMetric` through to Ruby- and Java- metrics APIs
- [x] migrate all internal timers
- [x] provide opt-in live-tracking implementation of `TimerMetric`
- [x] refactor metrics-related code from `QueueReadClient` up to `WorkerLoop` (or elsewhere)
- ~[ ] validate performance of live-tracking `TimerMetric`, and make it the default~ _[deferred: #14809]_

## How to test this PR locally

The metrics underneath `duration_in_millis` for filters, outputs, codecs, pipelines, and process have all been moved to the new timers API, so they should work with net zero change. Similarly `queue_push_duration_in_millis` for inputs, the pipeline, and the process have been moved.

I enjoy using a bursty pipeline:
~~~
input { java_generator { threads => 16 } }

filter {
  ruby {
    init => 'require "prime"'
    code => "
      if Prime.prime?(Time.now.min) && (Random.rand(100) <= 1)
        sleep(2.0 ** Random.rand(4.0))
      end
    "
  }
}

output { sink { } }
~~~

And observing the Node Stats API to see these metrics (and the flow metrics that rely on them) changing in real-time:

~~~
curl -XGET 'localhost:9600/_node/stats'
~~~
